### ISSUE TITLE
SSL cipher suite list includes some insecure options

### ISSUE THREAD

**Issue by Synchro** [Sequel-Ace/Sequel-Ace#164]
The list of supported SSL cipher suites includes some options that are considered broken or at best inadvisable:

![image](https://user-images.githubusercontent.com/81561/86491027-c30f5880-bd69-11ea-8920-6dd11797c2c5.png)

In particular anything using RC4, CBC, MD5, SHA-1. While this may not present a significant risk because SA is a client rather than a server, It might still be better to disable known-bad options by default so that they need to be explicitly enabled by users.

**Comment by jamesstout** [comment]
@gboudreau Do you know if there's a way to tell which cipher is used during the connection?

**Comment by gboudreau** [comment]
SHOW STATUS LIKE 'Ssl_cipher';

**Comment by gboudreau** [comment]
Of note: since we updated libmysqlclient to 8, this list might be out of date (might contain ciphers not even supported by libmysqlclient, and might be missing new ones that are supported now, and weren't before).

**Comment by gboudreau** [comment]
Best documentation page I could find on ciphers: https://dev.mysql.com/doc/refman/8.0/en/encrypted-connection-protocols-ciphers.html#encrypted-connection-cipher-configuration

Strings found in the libmycqlclient.a we use:

```
ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256
ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:DHE-RSA-AES128-SHA256:DHE-DSS-AES128-SHA256:DHE-DSS-AES256-GCM-SHA384:DHE-RSA-AES256-SHA256:DHE-DSS-AES256-SHA256:DHE-RSA-AES256-GCM-SHA384
DH-DSS-AES128-GCM-SHA256:ECDH-ECDSA-AES128-GCM-SHA256:DH-DSS-AES256-GCM-SHA384:ECDH-ECDSA-AES256-GCM-SHA384:DH-DSS-AES128-SHA256:ECDH-ECDSA-AES128-SHA256:DH-DSS-AES256-SHA256:ECDH-ECDSA-AES256-SHA384:DH-RSA-AES128-GCM-SHA256:ECDH-RSA-AES128-GCM-SHA256:DH-RSA-AES256-GCM-SHA384:ECDH-RSA-AES256-GCM-SHA384:DH-RSA-AES128-SHA256:ECDH-RSA-AES128-SHA256:DH-RSA-AES256-SHA256:ECDH-RSA-AES256-SHA384
ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE-DSS-AES128-SHA:DHE-RSA-AES128-SHA:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:DH-DSS-AES128-SHA:ECDH-ECDSA-AES128-SHA:AES256-SHA:DH-DSS-AES256-SHA:ECDH-ECDSA-AES256-SHA:DH-RSA-AES128-SHA:ECDH-RSA-AES128-SHA:DH-RSA-AES256-SHA:ECDH-RSA-AES256-SHA:CAMELLIA256-SHA:CAMELLIA128-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA

!aNULL:!eNULL:!EXPORT:!LOW:!MD5:!DES:!RC2:!RC4:!PSK:!DES-CBC3-SHA:!DHE-DSS-DES-CBC3-SHA:!DHE-RSA-DES-CBC3-SHA:!ECDH-RSA-DES-CBC3-SHA:!ECDH-ECDSA-DES-CBC3-SHA:!ECDHE-RSA-DES-CBC3-SHA:!ECDHE-ECDSA-DES-CBC3-SHA:!DH-RSA-DES-CBC3-SHA:!DH-DSS-DES-CBC3-SHA
```

**Comment by sapphirecat** [comment]
FWIW, I did some experiments against my AWS RDS server (MySQL 8.0.20).

- With all ciphers disabled in the Sequel Ace UI, the connection fails.
- With any AES+SHA combo enabled, the connection succeeds, reporting `TLS_AES_256_GCM_SHA384` as Ssl\_cipher.
- With all other ciphers, the connection fails.

Specifically, `DHE-RSA-AES128-SHA`, `AES128-SHA`, and the AES256 versions of them succeed. The connection fails with RC4, DES, or RMD.  (AES+SHA DSS succeeded, but I believe guidance on using DSS these days is "never do that.")

**Comment by Jason-Morcos** [comment]
@sapphirecat Thanks for the investigation! I think in the short term let's focus on adding missing ciphers as requested and only removing ones not supported in mySQL >= 5.6. It appears mySQL will reject insecure ciphers so we don't have to worry that much about an insecure cipher being attempted, as I understand it at least.

**Comment by wedi** [comment]
From what I understand, it's optional in libmysqlclient to set the allowed ciphers.

I'd suggest to disable all ciphers, let libmysqlclient choose and optionally display the actually used cipher somewhere.
If the literate, constrained or snakeoil loving user wants to use something else let him activate the ciphers he wants. This would yield a reasonable, always up-to-date cipher selection for the default user and eliminate any discussion about default order and how (un-)secure the selected ciphers are.

**Comment by Jason-Morcos** [comment]
Should just need to alter the `defaultSSLCipherList` (line 519+ of SPNetworkPreferencePane.m). The hard part here is knowing what ciphers should be in this list.

**Comment by Jason-Morcos** [comment]
Relating to Cipher security, @Synchro do you believe this PR would close this hole? Does it add any new ciphers which should not be allowed? https://github.com/Sequel-Ace/Sequel-Ace/pull/428

**Comment by Synchro** [comment]
I'm not clear whether this list is of the suites that you're providing support for (disabled or not), or just the ones that are enabled.

It's good that all the RC4, MD5, and CBC suites are gone.

From that list I would recommend disabling all that start with `DH-` or `ECDH-` (no forward secrecy) and end with `-SHA` (weak SHA-1 signature). Ones that don't specify a KX (like `AES128-SHA256`) use RSA for KX, which is also considered a poor choice, and has been removed altogether in TLSv1.3. I'm not sure about DSS or ECDSA suites. If there is any order to this list, put the `GCM` ones first.

My only concern is that these changes might prevent people from connecting to legacy systems – in which case they should be disabled by default, but allowed to be explicitly enabled if needed.

**Comment by Jason-Morcos** [comment]
> I'm not clear whether this list is of the suites that you're providing support for (disabled or not), or just the ones that are enabled.
The list in that PR is all the supported suites. Those above the separator line thing are the ones that are enabled by default!

> 
> It's good that all the RC4, MD5, and CBC suites are gone.
> 
> From that list I would recommend disabling all that start with `DH-` or `ECDH-` (no forward secrecy) and end with `-SHA` (weak SHA-1 signature). Ones that don't specify a KX (like `AES128-SHA256`) use RSA for KX, which is also considered a poor choice, and has been removed altogether in TLSv1.3. I'm not sure about DSS or ECDSA suites. 
Made these changes!

> If there is any order to this list, put the `GCM` ones first.
There is order! Good tip, will make this change.

> My only concern is that these changes might prevent people from connecting to legacy systems – in which case they should be disabled by default, but allowed to be explicitly enabled if needed.
Agreed. We should include insecure ciphers, but they should be disabled by default


Please take a look at the list again - I updated it based on your recommendations

**Comment by Jason-Morcos** [comment]
Changes to ciphers will be included in the 2.3.0 release, the first beta of which is now available!

https://github.com/Sequel-Ace/Sequel-Ace/releases
